### PR TITLE
JIT: fix invariant analysis for cloning

### LINKED ISSUE #61040: Loop cloning misses definitions for fields of promoted structs
Reproduction:
```cs
using System.Runtime.CompilerServices;

Problem(default);

[MethodImpl(MethodImplOptions.NoInlining)]
static void JitUse<T>(T arg) { }

[MethodImpl(MethodImplOptions.NoInlining | MethodImplOptions.AggressiveOptimization)]
static void Problem(ArrayWrapper a)
{
    a = GetArrayLong();

    JitUse(a);
    JitUse(a);

    for (int i = 0; i < 10000; i++)
    {
        a = GetArray();
        JitUse(a.Array[i]);
    }
}

[MethodImpl(MethodImplOptions.NoInlining)]
static ArrayWrapper GetArray() => new() { Array = new int[0] };

[MethodImpl(MethodImplOptions.NoInlining)]
static ArrayWrapper GetArrayLong() => new() { Array = new int[10000] };

struct ArrayWrapper
{
    public int[] Array;
}
```
We expect to get an `IndexOutOfRangeException` here, but instead get an AV.

The cause is that loop cloning thinks that the promoted struct's field is invariant in the loop and deletes the bounds check. The reason it thinks it is invariant is because this is a special case in block morphing, where we leave an independently promoted struct [alone](https://github.com/dotnet/runtime/blob/f9a24499a1168d391cbf61ab57470a5a974845bb/src/coreclr/jit/morphblock.cpp#L693), and it looks like `optIsVarAssgCB` does not account for this. I thus conjecture this will also reproduce for the multi-reg assignment form.

Found while experimenting with morphing of the "one field" case into `ASG(LCL_VAR(promoted field), BITCAST(field type(CALL)))`.

@dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Fix some cases where the JIT was not sufficiently careful in verifing that
operands in a loop were suitably invariant.

Closes #61040.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib 

This is intended to have fairly minimal diffs (currently around 150 methods). Diffs will show both size decreases (because we stop cloning in some places) and size increases (because we lose a cloning condition and so either fail to remove bounds checks or fail to prove the slow loop is unreachable).

Spot checking showed the diffs to be reasonable; not clear if any of them were actual bugs or just places we though we knew something that we didn't actually know.

Some of the analysis here can be sharpened (eg to address #70100, and to improve the use of bit vectors, and possibly to allow some address-exposed cases).  I will work on that in a follow-on change.

File: src/coreclr/jit/optimizer.cpp [hunk 1]
```diff
@@ -909,20 +907,46 @@ bool Compiler::optCheckIterInLoopTest(
...
     else if (limitOp->gtOper == GT_ARR_LENGTH)
     {
-        optLoopTable[loopInd].lpFlags |= LPFLG_ARRLEN_LIMIT;
+        // See if limit array is a loop invariant
+        //
+        GenTree* const array = limitOp->AsArrLen()->ArrRef();
+
+        if (array->OperIs(GT_LCL_VAR))
+        {
+            if (!optIsVarAssgLoop(loopInd, array->AsLclVarCommon()->GetLclNum()))
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:931]: Do you want to add JITDUMP if this fails, like you did above? (Same for if it's not a GT_LCL_VAR?)

File: src/coreclr/jit/optimizer.cpp [hunk 2]
```diff
@@ -909,20 +907,46 @@ bool Compiler::optCheckIterInLoopTest(
...
+        }
     }
     else
     {
+        // Not clear why we return false here. We know the loop exit test
+        // is comparing iterVar to something; we just don't know what.
+        //
+        // No real difference between this case and the cases above where
+        // we don't set some kind of limit flag.
+        //
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:944]: I think the caller expects one of the `LPFLG_*_LIMIT` flags to be set if this function returns true; you are currently not doing that for, e.g., cases where `GT_ARR_LENGTH` fails.

File: src/coreclr/jit/optimizer.cpp [hunk 3]
```diff
@@ -977,7 +1001,7 @@ unsigned Compiler::optIsLoopIncrTree(GenTree* incr)
 // optComputeIterInfo: Check tree is loop increment of a lcl that is loop-invariant.
 //
 // Arguments:
-//      from, to    - are blocks (beg, end) which are part of the loop.
+//      lnum        - loop in question
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:1004]: You didn't change the actual function signature.

It's also a little weird to pass around the loop index when this function is called as part of populating the loop info, so the loop table for this loop isn't complete (I would generally assume that if given a loop index, all the info about the loop at that index was valid).

File: src/coreclr/jit/optimizer.cpp [hunk 4]
```diff
@@ -5736,38 +5806,97 @@ bool Compiler::optIsVarAssigned(BasicBlock* beg, BasicBlock* end, GenTree* skip,
...
+            if (parentVar < lclMAX_ALLSET_TRACKED)
+            {
+                JITDUMP("optIsVarAssgLoop: V%02u promoted, also checking V%02u\n", var, parentVar);
+                AllVarSetOps::AddElemD(this, vs, parentVar);
+            }
+            else
+            {
+                // Parent var index is too large, assume the worst.
+                //
+                return false;
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:5853]: Shouldn't this return `true`?

File: src/coreclr/jit/optimizer.cpp [hunk 5]
```diff
@@ -5736,38 +5806,97 @@ bool Compiler::optIsVarAssigned(BasicBlock* beg, BasicBlock* end, GenTree* skip,
...
+            }
+        }
+
         return optIsSetAssgLoop(lnum, vs) != 0;
     }
     else
     {
+        if (varDsc->lvIsStructField)
+        {
+            return false;
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:5863]: Shouldn't this return `true`?

**Review by BruceForstall — CHANGES_REQUESTED** [review]

File: src/coreclr/jit/optimizer.cpp [hunk 6]
```diff
@@ -5671,20 +5726,20 @@ Compiler::fgWalkResult Compiler::optIsVarAssgCB(GenTree** pTree, fgWalkData* dat
...
+            // We can't track every field of every var. Moreover, indirections
+            // may access different parts of the var as different (but
+            // overlapping) fields. So just treat them as indirect accesses
+            //
             // unsigned    lclNum = dest->AsLclFld()->GetLclNum();
             // noway_assert(lvaTable[lclNum].lvAddrTaken);
-
+            //
             varRefKinds refs = varTypeIsGC(tree->TypeGet()) ? VR_IND_REF : VR_IND_SCL;
             desc->ivaMaskInd = varRefKinds(desc->ivaMaskInd | refs);
```
Comment (SingleAccretion) [inline src/coreclr/jit/optimizer.cpp:5737]: Would it be out of the scope for this change to also fix cases where "treating `LCL_FLD` conservatively as indirect access" is not enough, and where we can see `OBJ/BLK`-based stores?

<details>
<summary>Test cases that expose the two issues</summary>

```cs
using System;
using System.Runtime.CompilerServices;

unsafe class LoopLocalInvariance
{
    private const int ArrLen = 10;

    public static int Main()
    {
        int[] arr = new int[ArrLen];

        try
        {
            ProblemWithBlkAsg(arr, ArrLen);
            return 101;
        }
        catch (IndexOutOfRangeException) { }

        try
        {
            ProblemWithLclFldAsg(arr, ArrLen);
            return 102;
        }
        catch (IndexOutOfRangeException) { }

        return 100;
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void ProblemWithBlkAsg(int[] arr, int idx)
    {
        for (int i = 0; i < ArrLen; i++)
        {
            Unsafe.InitBlock(&i, (byte)idx, 1);
            arr[i] = 0;
        }
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void ProblemWithLclFldAsg(int[] arr, int idx)
    {
        for (int i = 0; i < ArrLen; i++)
        {
            *(byte*)&i = (byte)idx;
            arr[i] = 0;
        }
    }
}
```
</details>

File: src/coreclr/jit/optimizer.cpp (hunk 4 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/optimizer.cpp:5853]: Oops. Yes.

File: src/coreclr/jit/optimizer.cpp (hunk 6 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/optimizer.cpp:5737]: I'll take a look, sure.

I have follow on changes where I want to enable more cloning (an in more cases) so would be good to fix all the issues we know of with the analysis.

**Review by BruceForstall — APPROVED** [review]

**Review by SingleAccretion — APPROVED** [review]

**Comment by AndyAyersMS** [comment]
Left off adding a one test case. Will tack it onto a subsequent PR.

**Pull request merged**